TrkFilters/src/MultiHelixFilter_module.cc

```diff
@@ -0,0 +1,589 @@
+  }
+
+  //--------------------------------------------------------------------------------------
+  bool MultiHelixFilter::checkHelixList(std::vector<const HelixSeed*> helices) {
+    // check the collection can pass the cuts
+    if(helices.empty() || helices.size() > _helixCuts.size()) return false;
+
+
+    // check if each helix passes the corresponding cut set
+    double min_time(helices[0]->t0().t0()), max_time(helices[0]->t0().t0()); // to get the time window of the helices
+    for(size_t index = 0; index < helices.size(); ++index) {
+      if(!_helixCuts[index].applyCuts(helices[index])) return false;
+      min_time = std::min(min_time, helices[index]->t0().t0());
+      max_time = std::max(max_time, helices[index]->t0().t0());
+    }
+
+    // check the helices are contained in the time window
+    if(_timeWindow > 0. && max_time - min_time > _timeWindow) return false;
+
+    // check that all helices pass the overlap check
+    const bool do_overlap = _maxOverlap > 0. && _maxOverlap < 1.;
+    const bool do_arc     = _minArcGap > 0.;
+    if(do_overlap || do_arc) {
+      for(size_t i = 0; i < helices.size() - 1; ++i) {
+        for(size_t j = i+1; j < helices.size(); ++j) {
+          const HelixSeed* h1 = helices[i];
+          const HelixSeed* h2 = helices[j];
+          if(do_overlap && evaluateOverlap(h1,h2) > _maxOverlap) return false;
+          // check if it they share a side of the helix circle
+          if(do_arc && evaluateArcGap(h1, h2) < _minArcGap) return false;
+        }
+      }
+    }
+
+    if(_debugLevel > 0) std::cout << ">>> [MultiHelixFilter::" << __func__ << "::" << moduleDescription().moduleLabel() << "]"
+                                  << " Accepted a helix set\n";
+
+    // helix set is accepted
+    return true;
+  }
+
+  //--------------------------------------------------------------------------------------
+  void MultiHelixFilter::doPermutations(const std::vector<size_t>& given, const HelixSeedCollection* hcol, std::set<size_t>& accepted) {
+    const size_t nhelices = hcol->size();
+    if(nhelices < _minNHelices) return; // can't pass the requirements
+
+    if(_debugLevel > 2) {
+      std::cout << __func__ << ": Level " << given.size() << ": input set = {";
+      for(size_t index = 0; index < given.size(); ++index) {
+        if(index > 0) std::cout << ", ";
+        std::cout << given[index];
+      }
+      std::cout << "}, N(helices) = " << nhelices << std::endl;
+    }
+
+    // if more than two free indices, continue down
+    if(given.size() < _minNHelices) {
+      for(size_t index = 0; index < nhelices; ++index) {
+        if(find(given.begin(), given.end(), index) != given.end()) continue; //already in the set
+        auto next_level = given;
+        next_level.push_back(index);
+        doPermutations(next_level, hcol, accepted);
+      }
+      return;
+    }
+
+    // at the base level where all indices are fixed
+    std::vector<const HelixSeed*> helices;
```

Comment:
Im curious why you use pointers to store the helix references here, but everywhere else you use indices.